src/scenic/core/object_types.py

```diff
@@ -3,666 +3,1760 @@
+            # Can't get a bound on one or more dimensions, abort
+            return 0
+
+        min_bounds = np.array([min_width, min_length, min_height])
+        max_bounds = np.array([max_width, max_length, max_height])
+
+        # Extract a list of possible shapes
+        if isinstance(self.shape, Shape):
+            shapes = [self.shape]
+        elif hasattr(self.shape, "options"):
+            if all(isinstance(shape, Shape) for shape in self.shape.options):
+                shapes = list(self.shape.options)
+            else:
+                # Something we don't recognize, abort
+                return 0
+
+        # Check that all possible shapes contain the origin
+        if not all(shape.containsCenter for shape in shapes):
+            # One or more shapes has inradius 0
+            return 0
+
+        # Get the inradius for each shape with the min and max bounds
+        min_distances = [MeshVolumeRegion(mesh=shape.mesh, dimensions=min_bounds).inradius for shape in shapes]
+        max_distances = [MeshVolumeRegion(mesh=shape.mesh, dimensions=max_bounds).inradius for shape in shapes]
+
+        distance_range = (min(min_distances), max(max_distances))
+
+        return InradiusHelper(support=distance_range)
+
+    @cached_property
+    def surface(self):
+        """A region containing the entire surface of this object"""
+        return self.occupiedSpace.getSurfaceRegion()
+
+    @cached_property
+    def onSurface(self):
+        """The surface used by the ``on`` specifier.
+
+        This region is used to sample position when
+        another object is placed ``on`` this object. By default
+        the top surface of this object (`topSurface`), but can
+        be overwritten by subclasses.
+        """
+        return self.topSurface
+
+    @cached_property
+    def rightSurface(self):
+        """A region containing the right surface of this object"""
+        return defaultSideSurface(self.occupiedSpace, dimension=0, \
+            positive=True, thresholds=self.sideComponentThresholds)
+
+    @cached_property
+    def leftSurface(self):
+        """A region containing the left surface of this object"""
+        return defaultSideSurface(self.occupiedSpace, dimension=0, \
+            positive=False, thresholds=self.sideComponentThresholds)
+
+    @cached_property
+    def frontSurface(self):
+        """A region containing the front surface of this object"""
+        return defaultSideSurface(self.occupiedSpace, dimension=1, \
+            positive=True, thresholds=self.sideComponentThresholds)
+
+    @cached_property
+    def backSurface(self):
+        """A region containing the back surface of this object"""
+        return defaultSideSurface(self.occupiedSpace, dimension=1, \
+            positive=False, thresholds=self.sideComponentThresholds)
+
+    @cached_property
+    def topSurface(self):
+        """A region containing the top surface of this object"""
```

Comment:
Since this one is used for `onSurface` by default, I'd have it come first out of the 6 and include a cross-reference to `defaultSideSurface` so people can look up how these surfaces are defined.